cartography/graph/loader.py

```diff
@@ -0,0 +1,144 @@
+    :param batch_size: Optional: the number of nodes to write in a single batch operation.
+    Default = 10000.
+    :return: None.
+    """
+    if batch_size < 1:
+        raise ValueError("batch_size cannot be less than 1.")
+
+    ingest_query = build_node_ingestion_query(node_label, node_property_map)
+
+    chunk = node_data_list[:batch_size]
+    cursor = 0
+    while len(chunk) > 0:
+        query_args = {
+            'UpdateTag': update_tag,
+            'DictList': chunk,
+            'NodeLabel': node_label,  # used for debug logs
+        }
+        neo4j_session.write_transaction(_merge_chunk, ingest_query, query_args)
+        cursor += batch_size
+        chunk = node_data_list[cursor:cursor + batch_size]
+
+
+def _merge_chunk(tx: neo4j.Transaction, ingest_query: str, query_args: Dict) -> None:
+    """
+    `unit_of_work` function for `neo4j_session.write_transaction()`.
+    """
+    result: neo4j.BoltStatementResult = tx.run(ingest_query, **query_args)
+    summary: neo4j.BoltStatementResultSummary = result.consume()
+
+    if 'NodeLabel' in query_args:
+        logger.debug(f"Created {summary.counters.nodes_created} nodes with label '{query_args['NodeLabel']}'.")
+    if 'RelLabel' in query_args:
+        logger.debug(
+            f"Created {summary.counters.relationships_created} relationships with label '{query_args['RelLabel']}'.",
+        )
+
+
+def merge_relationships(
+    neo4j_session: neo4j.Session,
+    node_label_a: str, search_property_a: str, dict_key_a: str,
+    node_label_b: str, search_property_b: str, dict_key_b: str,
+    relationship_label: str, rel_mapping_list: List[Dict],
+    update_tag: int,
+    rel_property_map: Dict[str, str] = None,
+    batch_size: int = 10000,
+) -> None:
+    """
+    Writes relationships from nodes A to B using the `rel_mapping_list`. Uses the
+    UNWIND pattern in batches with just one MERGE per statement for performance and
+    reliability. Follows guidelines described in
+    - https://neo4j.com/blog/bulk-data-import-neo4j-3-0/
+    - https://dzone.com/articles/tips-for-fast-batch-updates-of-graph-structures-wi
+
+    The query run looks like
+
+    UNWIND {RelMappingList} AS item
+        MATCH (a:`node_label_a`{`search_property_a`:item.`dict_key_a`})
+        MATCH (b:`node_label_b`{`search_property_b`:item.`dict_key_b})
+        MERGE (a)-[r:`rel_label`]->(b)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {UpdateTag},
+            ... <optionally expands the relationship property map too> ...
+
+    :param neo4j_session: The Neo4j transaction object.
+    :param node_label_a: The label of node A.
+    :param search_property_a: the search key to use to search the graph to find node A. For
+    performance, this should be an indexed property. If your graph is large, querying on
+    non-indexed properties can cause your syncs to take **days** to run!
+    :param dict_key_a: The dict key on what value of `search_property_a` to search for.
```

Comment:
i think search_property and dict_key are confusing given that we've referred to properties as fields before. what about `field_name` and `field_value`?